### PR TITLE
[TNB-18] Game level configuration

### PR DESCRIPTION
**Pull request by Squidfishrl**
## What's here?
- Add `LevelRepository` (reads json config files and converts them to `GameSettings`)
- Start games with level specific settings 

## Tickets involved
#18

### CODE REVIEW

File: app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelMapper.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        }
+
+        return Result.Success(
+            GameSettings(
+                recallsBack = settings.recallsBack,
+                totalRounds = settings.totalRounds,
+                millisPerRound = settings.millisPerRound,
+                repeatChance = settings.repeatChance,
+            ),
+        )
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelMapper.kt]: Maybe combine at least the guards into a `when` expression

File: app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelRepository.kt [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+/**
+ * Get level configuration for [Game]s.
+ *
+ * @param[context] Application context
+ * @param[autoInitCache] When true the content of all config files is automatically cached. See
+ * [initializeCache]
+ */
+class LevelRepository(private val context: Context, private val autoInitCache: Boolean = false) :
+    SearchRepository<GameSettings, LevelData> {
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelRepository.kt:30]: Format the params on new lines instead of super type being on a new line

File: app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelRepository.kt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    override suspend fun insert(data: GameSettings): Result<Unit, Error> {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun delete(data: GameSettings): Result<Unit, Error> {
+        TODO("Not yet implemented")
+    }
+}
+
+sealed class CacheInitializationState {
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelRepository.kt]: Seems like it can be an interface

File: app/src/main/java/org/squidfish/tuplenback/di/ViewModelModule.kt [hunk 4]
```diff
@@ -3,13 +3,16 @@ package org.squidfish.tuplenback.di
 import org.koin.android.annotation.KoinViewModel
 import org.koin.core.annotation.Module
 import org.squidfish.tuplenback.data.game.GameRepository
+import org.squidfish.tuplenback.data.game.levels.LevelRepository
 import org.squidfish.tuplenback.models.GameViewModel
 import org.squidfish.tuplenback.presentation.screen.gameselection.GameSelectionViewModel
 
 @Module
 class ViewModelModule {
+    // FIXME: ktlint disagrees with every other way to write this
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/di/ViewModelModule.kt:12]: Will fix in another ticket, it's not a blocker

File: app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt [hunk 5]
```diff
@@ -52,11 +65,20 @@ class GameViewModel(private val repository: RecentRepository<GameModel>) : ViewM
...
+                    val settings = when (val res = levelRepository.get(it.asLevelData)) {
+                        is Result.Error -> {
+                            Log.e(TAG, "Cannot initialize ViewModel. Level Repository error: $it - $res")
+                            return@launch
+                        }
+                        is Result.Success -> res.data
+                    }
+
+                    _level.value = Level(it.level, it.gameType, settings)
+                }
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt]: Can we offload this to the `onStart` of the state?

File: app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt [hunk 6]
```diff
@@ -115,18 +135,46 @@ class GameViewModel(private val repository: RecentRepository<GameModel>) : ViewM
...
+
+            gameEngines.clear()
+
+            initGameEngines().onError {
+                _gameLoadingState.value = GameLoadingState.Failiure(it)
+                return@launch
+            }
+            initGameState().onError {
+                _gameLoadingState.value = GameLoadingState.Failiure(it)
+                return@launch
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:168]: No need for the labels on the `return`s, there is nothing else in the function to run

File: app/src/main/java/org/squidfish/tuplenback/utils/DataErrors.kt
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/utils/DataErrors.kt:1]: I appreciate the generic naming of the error cases but can we move this our of the `utils` package and have them closer to their respective feature (though it would saty in domain because of their presence in domain signatures). That would most probably mean the `DataError` interface will stop being `sealed` but it seems like it is supposed to be extended with more errors anyway

File: app/src/main/java/org/squidfish/tuplenback/views/TupleNBackApp.kt [hunk 7]
```diff
@@ -105,12 +140,12 @@ fun TupleNBackApp(
...
                     },
                     onPlayRecent = {
-                        if (gameModel.game == null) {
+                        if (gameModel.level.value == null) {
                             Log.w(TAG, "Cannot play recent: No game played previously")
                             return@MainMenuScreen
                         }
 
-                        navController.navigate(ScreenDestination.GameScreen(game = null))
+                        navController.navigate(ScreenDestination.GameScreen(null, 0))
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/views/TupleNBackApp.kt:148]: When using constants, can we keep the labels for clarity? I don't know why detekt is not complaining but I am a little bit, I had to check line 111 to see what `0` means

File: app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+import org.squidfish.tuplenback.utils.Result
+import org.squidfish.tuplenback.utils.data
+
+@RunWith(RobolectricTestRunner::class)
+class LevelRepositoryTests {
+    private lateinit var realLevelRepository: LevelRepository
+    private lateinit var spiedLevelRepository: LevelRepository
+    private lateinit var context: Application
+
+    val singleGridLevelConfig = """
```
Comment (Stanislav04) [inline app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt]: `private`?

File: app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+                        "repeatChance": 20
+                    }
+                }
+            ]
+        }
+    """.trimIndent()
+
+    @Before
+    fun createLevelRepository() {
+        this.context = ApplicationProvider.getApplicationContext<Application>()
```
Comment (Stanislav04) [inline app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt]:
```suggestion
        context = ApplicationProvider.getApplicationContext<Application>()
```
I am not sure if we need to specify the generic parameter either

File: app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+
+    @After
+    fun tearDown() {
+        stopKoin()
+    }
+
+    @Test
+    fun `load level`() = runTest {
+        // Given
+        doReturn(Result.Success(singleGridLevelConfig)).`when`(spiedLevelRepository).readLevelData(Game.Grid)
```
Comment (Stanislav04) [inline app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt:61]: Okay, I do notice it's not the Mockito I have used before

File: app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+        val settings = spiedLevelRepository.get(LevelData(Game.Grid, 0))
+
+        // Then
+        assert(
+            settings.data == GameSettings(
+                recallsBack = 2,
+                totalRounds = 3,
+                millisPerRound = 1000,
+                repeatChance = 20,
+            ),
```
Comment (Stanislav04) [inline app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt]: Not that the trailing comma is that big of a deal, but I think there is an `asert` version that takes a lambda. Might be called `assertTrue` though

File: app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt
Comment (Stanislav04) [inline app/src/test/java/org/squidfish/tuplenback/data/model/LevelRepositoryTests.kt:1]: Nice stuff (I have not gone through the logical check if every corner case has been covered, I hope they are)

**Review by Stanislav04 — APPROVED** [review]
Maybe move some files into their respective layer's package, but otherwise great work

File: app/src/main/java/org/squidfish/tuplenback/utils/DataErrors.kt
Comment (Squidfishrl) [reply app/src/main/java/org/squidfish/tuplenback/utils/DataErrors.kt:1]: Would you mind elaborating? If my assumptions are correct, you're suggesting the dissolution of the utils package, that is, not merely moving it to the domain layer. Furthermore, how would, for example, `LevelDeserializationError` be closer to `LevelRepository`, given that the former would be in the domain layer whereas the latter in the data layer?

File: app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt (hunk 6 above)
Comment (Squidfishrl) [reply app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:168]: That doesn't seem to work. Apparently, you cannot have non-qualified return statements within lambda functions (which `viewModelScope.launch` takes), since there's compiler ambiguity from whence to return.

File: app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelRepository.kt (hunk 2 above)
Comment (Squidfishrl) [reply app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelRepository.kt:30]: ktlint continues complaining even so. The best I can do is the following, though I find myself preferring the super type on a new line:
```kt
class LevelRepository : SearchRepository<GameSettings, LevelData> {
    private val context: Context
    private val autoInitCache: Boolean

    constructor(context: Context, autoInitCache: Boolean = false) {
        this.context = context
        this.autoInitCache = autoInitCache
    }
```

Comment (Stanislav04) [reply app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelRepository.kt:30]: No need to write a secondary constructor because of linting issues

File: app/src/main/java/org/squidfish/tuplenback/utils/DataErrors.kt
Comment (Stanislav04) [reply app/src/main/java/org/squidfish/tuplenback/utils/DataErrors.kt:1]: `LevelDeserializationError` can be in the file where the `LevelRepository` interface is defined which is also part of the domain layer

File: app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelMapper.kt [hunk 12]
```diff
@@ -7,28 +7,19 @@ import org.squidfish.tuplenback.utils.Result
...
 val GameLevel.asGameSettings: Result<GameSettings, Error>
     get() {
-        if (settings.recallsBack < 1) {
-            return Result.Error(LevelDeserializationError.InvalidRecallsBack)
-        }
-
-        if (settings.totalRounds <= settings.recallsBack) {
-            return Result.Error(LevelDeserializationError.TooFewRounds)
-        }
+        return when {
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/data/game/levels/LevelMapper.kt]: `get() = when { .... }`

File: app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt [hunk 13]
```diff
@@ -63,26 +63,6 @@ class GameViewModel(
...
-                    }
-
-                    _level.value = Level(it.level, it.gameType, settings)
-                }
-            }.onError {
-                Log.e(TAG, "Cannot initialize ViewModel. Game Repository error: $it")
-            }
-        }
-    }
-
```
Comment (Stanislav04) [inline app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:63]: I see this is just deleted, do we not need it?

**Review by Stanislav04 — APPROVED** [review]

File: app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt (hunk 13 above)
Comment (Squidfishrl) [reply app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:63]: My mistake. It seems redundant but we do need it. In any case, it should be refactored in the future.
